src/FSharpVSPowerTools.Logic/Outlining/OutliningTagger.fs

```diff
@@ -0,0 +1,196 @@
+        | SynMemberDefn.Member (binding, _) -> yield visitBinding binding
+        | SynMemberDefn.LetBindings (bindings, _, _, _) ->
+            yield! Seq.map visitBinding bindings
+        | SynMemberDefn.Interface(tp, mmembers, _) ->
+            yield mkFileIndexRange d.Range.FileIndex tp.Range.End d.Range.End
+            match mmembers with
+            | Some members -> yield! Seq.collect visitSynMemberDefn members
+            | None -> ()
+        | SynMemberDefn.NestedType(td, _, _) ->
+            yield! visitTypeDefn td
+        | _ -> ()
+    }
+
+and visitTypeDefn (TypeDefn(componentInfo, objectModel, members, range)) =
+    seq {
+        yield mkFileIndexRange range.FileIndex componentInfo.Range.End range.End
+        match objectModel with
+        | ObjectModel(_, members, _) -> yield! Seq.collect visitSynMemberDefn members
+        | Simple _ -> yield! Seq.collect visitSynMemberDefn members
+    }
+
+let rec visitDeclaration (decl: SynModuleDecl) = 
+    seq {
+        match decl with
+        | SynModuleDecl.Let(_, bindings, _) ->
+            yield! Seq.map visitBinding bindings
+        | SynModuleDecl.Types(types, _) ->
+            yield! Seq.collect visitTypeDefn types
+        | SynModuleDecl.NestedModule(cmpInfo, decls, _, _) ->
+            yield mkFileIndexRange decl.Range.FileIndex cmpInfo.Range.End decl.Range.End
+            yield! Seq.collect visitDeclaration decls
+        | _ -> ()
+    }
+
+let visitModuleOrNamespace moduleOrNs =
+    seq {
+        let (SynModuleOrNamespace(_, _, decls, _, _, _, _)) = moduleOrNs
+        yield! Seq.collect visitDeclaration decls
+    }
+
+let visitAst tree =
+    seq {
+        match tree with
+        | ParsedInput.ImplFile(implFile) ->
+            let (ParsedImplFileInput(_, _, _, _, _, modules, _)) = implFile
+            yield! Seq.collect visitModuleOrNamespace modules
+        | _ -> ()
+    }
+
+type SnapshotPoint with
+    static member OfPos(snapshot: ITextSnapshot, p: pos) =
+        let (line, column) = Pos.toZ(p)
+        snapshot.GetLineFromLineNumber(line).Start.Add(column)
+
+type SnapshotSpan with
+    static member OfRange(snapshot: ITextSnapshot, r: range) =
+        SnapshotSpan(
+            SnapshotPoint.OfPos(snapshot, r.Start),
+            SnapshotPoint.OfPos(snapshot, r.End))
+
+type Tagger
+    ( buffer: ITextBuffer
+    , textDocument: ITextDocument
+    , serviceProvider : IServiceProvider
+    , projectFactory: ProjectFactory
+    , languageService: VSLanguageService) as self = 
+
+    let tagsChanged = Event<_,_>()
+    let mutable snapshotSpans : SnapshotSpan [] = [||]
+
+    let triggerUpdate newSnapshotSpans =
+        snapshotSpans <- newSnapshotSpans
+        tagsChanged.Trigger(
+            self,
+            SnapshotSpanEventArgs(
+                SnapshotSpan(
+                    buffer.CurrentSnapshot,
+                    0,
+                    buffer.CurrentSnapshot.Length - 1)))
+
+    let createTagSpan (ss: SnapshotSpan) =
```

Comment:
`ss` is a very confusing name for a thing with a 25 lines scope. What about `span` or `snapshotSpan`?